esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6590,5 +6591,104 @@ bool CWsDfuEx::onDFUFilePublish(IEspContext &context, IEspDFUFilePublishRequest
+        CFilePartSizeThread(const char *_lfn, IPartDescriptor &_part, bool _compressed)
+            : lfn(_lfn), part(_part), compressed(_compressed) {};
+
+        virtual int run() override
+        {
+            for (unsigned c = 0; c < part.numCopies(); c++)
+            {
+                RemoteFilename rfn;
+                part.getFilename(c, rfn);
+                try
+                {
+                    Owned<IFile> file = createIFile(rfn);
+                    if (compressed)
+                    {
+                        compressedSize = file->size();
+                        Owned<IFileIO> io = createCompressedFileReader(file);
+                        if (io)
+                            size = io->size();
+                    }
+                    else
+                    {
+                        size = file->size();
+                    }
+                }
+                catch (IException *e)
+                {
+                    VStringBuffer tmp("%s: ", lfn.get());
+                    rfn.getPath(tmp);
+                    EXCLOG(e, tmp.str());
+                    e->Release();
+
+                    size = (offset_t)-1;
+                    compressedSize = (offset_t)-1;
+                }
+                if ((size != (offset_t)-1) && (!compressed || (compressedSize != (offset_t)-1)))
+                    break;
+            }
+            return 0;
+        };
+
+        offset_t getSize() { return size; }
+        offset_t getCompressedSize() { return compressedSize; }
+    };
+
+    //Get the sizes and the compressed sizes for each file parts in parallel
+    CIArrayOf<CFilePartSizeThread> threads;
+    bool compressed = fileDesc->isCompressed();
+    Owned<IPartDescriptorIterator> pi = fileDesc->getIterator();
+    ForEach(*pi)
+    {
+        CFilePartSizeThread* thread = new CFilePartSizeThread(lfn, pi->query(), compressed);
+        threads.append(*thread);
+        thread->start();
+    }
+    ForEachItemIn(i, threads)
+        threads.item(i).join();
+
+    //Set the size and the compressed size for each file parts and calculate file compressed size.
+    offset_t fileCompressedSize = (offset_t)0;
+    ForEachItemIn(ii, threads)
+    {
+        CFilePartSizeThread& thread = threads.item(ii);
+        offset_t size = thread.getSize();
+        offset_t compressedSize = thread.getCompressedSize();
+        if ((size == (offset_t)-1) || (compressed && (compressedSize == (offset_t)-1)))
+        {
+            if (compressed)
+                fileCompressedSize = (offset_t)-1;
+            IERRLOG("Failed to get part size for file %s part %u", lfn, ii);
+            break;
+        }
+
+        IPartDescriptor *partDesc = fileDesc->queryPart(ii);
+        IPropertyTree &props = partDesc->queryProperties();
+        props.setPropInt64("@size", size);
+        if (compressed)
+        {
+            props.setPropInt64("@compressedSize", compressedSize);
+            fileCompressedSize += compressedSize;
```

Comment:
see other comment below, I _think_ top-level @compressedSize should be automatically populated (if so, no need to accumulate here), and if it is not, there may be a bug elsewhere.

Reply:
will need https://github.com/hpcc-systems/HPCC-Platform/pull/14836, but this code can be removed.